gef.py

```diff
@@ -5207,6 +5205,199 @@ def do_invoke(self, args: List) -> None:
         return
 
 
+class ExternalStructureManager:
+    class Structure:
+        def __init__(self, mod_path: pathlib.Path, struct_name: str) -> None:
+            self.module_path = mod_path
+            self.name = struct_name
+            self.class_type = self.__get_structure_class()
+            return
+
+        def __str__(self) -> str:
+            return self.name
+
+        def pprint(self):
+            res = []
+            for _name, _type in self.class_type._fields_:
+                _size = ctypes.sizeof(_type)
+                __name = Color.colorify(_name, gef.config["pcustom.structure_name"])
+                __type = Color.colorify(_type.__name__, gef.config["pcustom.structure_type"])
+                __size = Color.colorify(hex(_size), gef.config["pcustom.structure_size"])
+                __offset = Color.boldify(f"{getattr(self.class_type, _name).offset:04x}")
+                res.append (f"{__offset}   {__name:32s}   {__type:16s}  /* size={__size} */")
+            print(os.linesep.join(res))
+            return
+
+        def __get_structure_class(self) -> Type:
+            """Returns a tuple of (class, instance) if modname!classname exists"""
+            fpath = self.module_path
+            spec = importlib.util.spec_from_file_location(fpath.stem, fpath)
+            module = importlib.util.module_from_spec(spec)
+            sys.modules[fpath.stem] = module
+            spec.loader.exec_module(module)
+            _class = getattr(module, self.name)
+            return _class
+
+        def apply_at(self, address: int, depth: int = 0, max_depth: int = -1) -> None:
+            """Apply (recursively if possible) the structure format to the given address."""
+            if depth >= max_depth:
+                warn("maximum recursion level reached")
+                return
+
+            # read the data at the specified address
+            _structure = self.class_type()
+            _sizeof_structure = ctypes.sizeof(_structure)
+
+            try:
+                data = gef.memory.read(address, _sizeof_structure)
+            except gdb.MemoryError:
+                err(f"{' ' * depth}Cannot read memory {address:#x}")
+                return
+
+            # deserialize the data
+            length = min(len(data), _sizeof_structure)
+            ctypes.memmove(ctypes.addressof(_structure), data, length)
+
+            # pretty print all the fields (and call recursively if possible)
+            ptrsize = gef.arch.ptrsize
+            for field in _structure._fields_:
+                _name, _type = field
+                _value = getattr(_structure, _name)
+                _offset = getattr(self.class_type, _name).offset
+
+                if (ptrsize == 4 and _type is ctypes.c_uint32) \
+                or (ptrsize == 8 and _type is ctypes.c_uint64) \
```

Comment:
you can put all the predicates in a set of parens and then align under it for readability

Reply:
wdym?

Reply:
instead of
```python
if (ptrsize == 4 and _type is ctypes.c_uint32) \
or (ptrsize == 8 and _type is ctypes.c_uint64) \ ...
```

```python
if ((ptrsize == 4 and _type is ctypes.c_uint32)
    or (ptrsize == 8 and _type is ctypes.c_uint64) ...
```

note the extra parens on the outside, and lack of backslash. Strictly a readability thing.